Source/Core/VideoCommon/VertexShaderGen.cpp

```diff
@@ -222,60 +222,57 @@ ShaderCode GenerateVertexShaderCode(APIType api_type, const ShaderHostConfig& ho
+              "float4 P2 = " I_TRANSFORMMATRICES "[posidx + 2];\n");
     if ((uid_data->components & VB_HAS_NRMALL) != 0)
     {
       out.Write("int normidx = posidx & 31;\n"
-                "float3 N0 = " I_NORMALMATRICES "[normidx].xyz, N1 = " I_NORMALMATRICES
-                "[normidx+1].xyz, N2 = " I_NORMALMATRICES "[normidx+2].xyz;\n");
-    }
-
-    if ((uid_data->components & VB_HAS_NRM0) != 0)
-    {
-      out.Write("float3 _norm0 = normalize(float3(dot(N0, rawnorm0), dot(N1, rawnorm0), dot(N2, "
-                "rawnorm0)));\n");
-    }
-    if ((uid_data->components & VB_HAS_NRM1) != 0)
-    {
-      out.Write(
-          "float3 _norm1 = float3(dot(N0, rawnorm1), dot(N1, rawnorm1), dot(N2, rawnorm1));\n");
-    }
-    if ((uid_data->components & VB_HAS_NRM2) != 0)
-    {
-      out.Write(
-          "float3 _norm2 = float3(dot(N0, rawnorm2), dot(N1, rawnorm2), dot(N2, rawnorm2));\n");
+                "float3 N0 = " I_NORMALMATRICES "[normidx].xyz;\n"
+                "float3 N1 = " I_NORMALMATRICES "[normidx + 1].xyz;\n"
+                "float3 N2 = " I_NORMALMATRICES "[normidx + 2].xyz;\n");
     }
   }
   else
   {
-    out.Write("float4 pos = float4(dot(" I_POSNORMALMATRIX "[0], rawpos), dot(" I_POSNORMALMATRIX
-              "[1], rawpos), dot(" I_POSNORMALMATRIX "[2], rawpos), 1.0);\n");
-    if ((uid_data->components & VB_HAS_NRM0) != 0)
-    {
-      out.Write("float3 _norm0 = normalize(float3(dot(" I_POSNORMALMATRIX
-                "[3].xyz, rawnorm0), dot(" I_POSNORMALMATRIX
-                "[4].xyz, rawnorm0), dot(" I_POSNORMALMATRIX "[5].xyz, rawnorm0)));\n");
-    }
-    if ((uid_data->components & VB_HAS_NRM1) != 0)
-    {
-      out.Write("float3 _norm1 = float3(dot(" I_POSNORMALMATRIX
-                "[3].xyz, rawnorm1), dot(" I_POSNORMALMATRIX
-                "[4].xyz, rawnorm1), dot(" I_POSNORMALMATRIX "[5].xyz, rawnorm1));\n");
-    }
-    if ((uid_data->components & VB_HAS_NRM2) != 0)
+    // One shared matrix
+    out.Write("float4 P0 = " I_POSNORMALMATRIX "[0];\n"
+              "float4 P1 = " I_POSNORMALMATRIX "[1];\n"
+              "float4 P2 = " I_POSNORMALMATRIX "[2];\n");
+    if ((uid_data->components & VB_HAS_NRMALL) != 0)
     {
-      out.Write("float3 _norm2 = float3(dot(" I_POSNORMALMATRIX
-                "[3].xyz, rawnorm2), dot(" I_POSNORMALMATRIX
-                "[4].xyz, rawnorm2), dot(" I_POSNORMALMATRIX "[5].xyz, rawnorm2));\n");
+      out.Write("float3 N0 = " I_POSNORMALMATRIX "[3].xyz;\n"
+                "float3 N1 = " I_POSNORMALMATRIX "[4].xyz;\n"
+                "float3 N2 = " I_POSNORMALMATRIX "[5].xyz;\n");
     }
   }
 
-  if ((uid_data->components & VB_HAS_NRM0) == 0)
+  out.Write("float4 pos = float4(dot(P0, rawpos), dot(P1, rawpos), dot(P2, rawpos), 1.0);\n");
+  if ((uid_data->components & VB_HAS_NRM0) != 0)
+  {
+    if ((uid_data->components & VB_HAS_NRM1) == 0)
+      out.Write("float3 rawnorm1 = " I_CACHED_TANGENT ".xyz;\n");
```

Comment:
Calling these normals is a bit strange.  AFAIK a normal is typically a single vector that is perpendicular to the surface.

Maybe instead of `rawnorm1` we use `rawtangent`.  Same for `rawnorm2` to be `rawbinormal`.

I imagine we also can change the `_norm1` as well.  I'm not sure yet what the `N0`, `N1`, etc values are so I can't say whether that variable makes sense.

Reply:
`N0`, `N1`, and `N2` are the rows of the normal transformation matrix; `P0`, `P1`, and `P2` are similarly the rows of the position transformation matrix.

Renaming `rawnorm1` and `_norm1` could be a bit invasive, but I'm not opposed to doing it.  (It'd also make sense to rename `_norm0` to just `_norm`.)  I think this naming scheme mostly comes from the vertex loader code referring to a normal count (either 1 or 3), and in turn Nintendo somewhat treating things the same way (though it's either `N` or `NBT`).

Reply:
There's also this (for D3D):

https://github.com/dolphin-emu/dolphin/blob/2e01dc0c82b5f73215e3425ab45ff6c7ee50b389/Source/Core/VideoCommon/VertexShaderGen.cpp#L171-L177

[D3D has dedicated semantics](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-semantics) for `NORMAL`/`TANGENT`/`BINORMAL`, though it doesn't explain what they do differently.  I'll switch to those too.

Reply:
Thanks pokechu, I think that'll make it easier for anyone who is more familiar with graphics logic to understand what is going on.  I might end up using these for an enhancement type feature in the near future.

I'm still curious why `N0`, `N1`, `N2` forms a separate matrix from the normal/binorma/tangent.  I'm not a graphics guru but I just find it surprising, I wonder what its purpose was.

Reply:
I wonder if it's to change the space of the vectors.  At least it looks like the normal/binormal/tangent are being orthnormalized to the new space?  (been a while since I've done that so I might be wrong)

Reply:
`N0`/`N1`/`N2` are rows of the single matrix that is used for transforming the normal, tangent, and binormal vectors, for transforming from the model space to the view space (or I guess tangent space?).  See `GX_LoadNrmMtxImm`.  It's usually the inverse transpose of the position matrix.

I'm not sure why we don't use a GLSL or HLSL matrix type for this, but perhaps there's some kind of compatibility reason?  I don't think matrices are used in the shaders at all currently (at least, I couldn't find any wrappers for them for compatibility between GLSL and HLSL).

Reply:
I've renamed them (and also confirmed that the shaders build successfully on all backends for both specialized and ubershaders and with/without per-pixel lighting).

Reply:
It would be nice if we could use matrix types for P0/P1/P2 and N0/N1/N2.

Would improve readability, and I don't think there is any limitation preventing it.

Reply:
I tried using HLSL's `float3x3` like this:

<details><summary>Patch, D3D only</summary>

```patch
diff --git a/Source/Core/VideoCommon/VertexShaderGen.cpp b/Source/Core/VideoCommon/VertexShaderGen.cpp
index dec2985ae8..89c3ffe43d 100644
--- a/Source/Core/VideoCommon/VertexShaderGen.cpp
+++ b/Source/Core/VideoCommon/VertexShaderGen.cpp
@@ -232,7 +232,8 @@ ShaderCode GenerateVertexShaderCode(APIType api_type, const ShaderHostConfig& ho
       out.Write("int normidx = posidx & 31;\n"
                 "float3 N0 = " I_NORMALMATRICES "[normidx].xyz;\n"
                 "float3 N1 = " I_NORMALMATRICES "[normidx + 1].xyz;\n"
-                "float3 N2 = " I_NORMALMATRICES "[normidx + 2].xyz;\n");
+                "float3 N2 = " I_NORMALMATRICES "[normidx + 2].xyz;\n"
+                "float3x3 Nmtx = float3x3(N0, N1, N2);\n");
     }
   }
   else
@@ -245,7 +246,8 @@ ShaderCode GenerateVertexShaderCode(APIType api_type, const ShaderHostConfig& ho
     {
       out.Write("float3 N0 = " I_POSNORMALMATRIX "[3].xyz;\n"
                 "float3 N1 = " I_POSNORMALMATRIX "[4].xyz;\n"
-                "float3 N2 = " I_POSNORMALMATRIX "[5].xyz;\n");
+                "float3 N2 = " I_POSNORMALMATRIX "[5].xyz;\n"
+                "float3x3 Nmtx = float3x3(N0, N1, N2);\n");
     }
   }
 
@@ -262,12 +264,9 @@ ShaderCode GenerateVertexShaderCode(APIType api_type, const ShaderHostConfig& ho
     // By normalising the first transformed normal (which is used by lighting calculations and needs
     // to be unit length), the same transform matrix can do double duty, scaling for emboss mapping,
     // and not scaling for lighting.
-    out.Write("float3 _normal = normalize(float3(dot(N0, rawnormal), dot(N1, rawnormal), dot(N2, "
-              "rawnormal)));\n"
-              "float3 _tangent = float3(dot(N0, rawtangent), dot(N1, rawtangent), dot(N2, "
-              "rawtangent));\n"
-              "float3 _binormal = float3(dot(N0, rawbinormal), dot(N1, rawbinormal), dot(N2, "
-              "rawbinormal));\n");
+    out.Write("float3 _normal = mul(Nmtx, rawnormal);\n"
+              "float3 _tangent = mul(Nmtx, rawtangent);\n"
+              "float3 _binormal = mul(Nmtx, rawbinormal);\n");
   }
   else
   {
```

</details>

but this doesn't work (it gives the wrong output, and renderdoc crashes when it gets to the `Nmtx` declaration; I get the wrong output with both `mul(Nmtx, rawnormal)` and `mul(rawnormal, Nmtx)` so I don't think this is a row-major vs column-major thing).

Reply:
This looks much better @Pokechu22 .  Thank you so much!

I think a matrix would help in the sense that we could then have one name:

`P0`, `P1`, `P2` -> `object_to_eye_space`
`N0`, `N1`, `N2` -> `object_to_tangent_space`

I'm just guessing at the conversion.

Reply:
You might need to set `#pragma pack_matrix( row_major )` for it to work as expected

Reply:
If we can't get matrices working, I think we should add a comment where the matrix multiply is done (with dot products) to make it clear to the reader that a matrix multiply is being implemented.

Reply:
And put the comment in the actual generated shader source... not just the code that generates it.

We kind of want those shaders to be readable too.

Reply:
I've done that for the first one (with the position matrix), but the same pattern is used throughout for both specialized shaders and ubershaders for normals, the projection matrix, texture coordinate transformation, and dual tex transform.  Adding it for each seems a bit messy, and is non-trivial for the texture coordinate ones due to how they're generated; it'd be nice to refactor it but I think it's out of scope for this PR.

Reply:
out-of-scope is fine with me.